File: vault_core/domain/impact.py
"""
Domain models for Impact Intelligence — assessment analytics and interventions.

Phase 1 — adds backend primitives for assessment intelligence
(schools, classes, learners, subjects, topics, assessments, questions,
mark entries, and interventions).

Single-tenant mode is unaffected; these tables remain empty until
school management is activated.
"""

from datetime import datetime
from typing import ClassVar, Optional

from vault_core.domain.base import ObjectModel


class ImpactSchool(ObjectModel):
    """A school or deploying organisation for Impact Intelligence."""

    table_name: ClassVar[str] = "impact_school"
    nullable_fields: ClassVar[set[str]] = {"district", "province", "school_type"}
    name: str
    district: Optional[str] = None
    province: Optional[str] = None
    school_type: Optional[str] = None  # "primary" | "secondary" | "tertiary"
    active: bool = True

    def __repr__(self) -> str:
        return f"ImpactSchool(id={self.id}, name={self.name})"


class ImpactClassGroup(ObjectModel):
    """A teaching group within a school, owned by a teacher."""

    table_name: ClassVar[str] = "impact_class_group"
    nullable_fields: ClassVar[set[str]] = {"teacher_name", "academic_year"}
    school_id: str
    name: str
    grade_level: Optional[str] = None
    academic_year: Optional[str] = None  # e.g. "2026"
    teacher_name: Optional[str] = None
    active: bool = True

    def __repr__(self) -> str:
        return f"ImpactClassGroup(id={self.id}, name={self.name}, school={self.school_id})"


class ImpactLearner(ObjectModel):
    """A learner enrolled in a class group."""

    table_name: ClassVar[str] = "impact_learner"
    nullable_fields: ClassVar[set[str]] = {"display_name"}
    school_id: str
    class_group_id: str
    learner_code: str  # school-assigned identifier
    display_name: Optional[str] = None
    status: str = "active"  # "active" | "inactive" | "transferred"

    def __repr__(self) -> str:
        return (
            f"ImpactLearner(id={self.id}, code={self.learner_code}, "
            f"class={self.class_group_id})"
        )


class ImpactSubject(ObjectModel):
    """A subject offered in the curriculum."""

    table_name: ClassVar[str] = "impact_subject"
    nullable_fields: ClassVar[set[str]] = {"level", "curriculum"}
    name: str
    level: Optional[str] = None  # e.g. "O-Level", "A-Level"
    curriculum: Optional[str] = None  # e.g. "ZIMSEC", "Cambridge"

    def __repr__(self) -> str:
        return f"ImpactSubject(id={self.id}, name={self.name})"


class ImpactTopic(ObjectModel):
    """A topic within a subject."""

    table_name: ClassVar[str] = "impact_topic"
    nullable_fields: ClassVar[set[str]] = {"strand", "syllabus_code"}
    subject_id: str
    name: str
    strand: Optional[str] = None
    syllabus_code: Optional[str] = None

    def __repr__(self) -> str:
        return f"ImpactTopic(id={self.id}, name={self.name}, subject={self.subject_id})"


class ImpactAssessment(ObjectModel):
    """An assessment (test, exam, quiz) for a class group."""

    table_name: ClassVar[str] = "impact_assessment"
    nullable_fields: ClassVar[set[str]] = {
        "term",
        "date_written",
        "pass_mark",
    }
    school_id: str
    class_group_id: str
    subject_id: str
    title: str
    assessment_type: str  # "test" | "exam" | "quiz" | "assignment"
    term: Optional[str] = None  # e.g. "Term 1"
    date_written: Optional[datetime] = None
    total_marks: int
    pass_mark: Optional[int] = None
    status: str = "draft"  # "draft" | "published" | "graded"

    def __repr__(self) -> str:
        return (
            f"ImpactAssessment(id={self.id}, title={self.title}, "
            f"type={self.assessment_type})"
        )


class ImpactAssessmentQuestion(ObjectModel):
    """A question within an assessment."""

    table_name: ClassVar[str] = "impact_assessment_question"
    nullable_fields: ClassVar[set[str]] = {"label", "difficulty", "topic_id"}
    assessment_id: str
    question_number: int
    label: Optional[str] = None  # e.g. "Q1", "Multiple Choice"
    max_marks: int
    topic_id: Optional[str] = None
    skill_type: str  # "knowledge" | "comprehension" | "application" | "analysis"
    difficulty: Optional[str] = None  # "easy" | "medium" | "hard"

    def __repr__(self) -> str:
        return (
            f"ImpactAssessmentQuestion(id={self.id}, "
            f"question={self.question_number}, assessment={self.assessment_id})"
        )


class ImpactMarkEntry(ObjectModel):
    """A mark entry for a learner on a specific question."""

    table_name: ClassVar[str] = "impact_mark_entry"
    assessment_id: str
    question_id: str
    learner_id: str
    score: float
    max_score: float

    def __repr__(self) -> str:
        return (
            f"ImpactMarkEntry(id={self.id}, score={self.score}, "
            f"learner={self.learner_id})"
        )


class ImpactIntervention(ObjectModel):
    """A recommended intervention based on assessment performance."""

    table_name: ClassVar[str] = "impact_intervention"
    nullable_fields: ClassVar[set[str]] = {"recommendation"}
    assessment_id: str
    class_group_id: str
    topic_id: str
    severity: str  # "low" | "medium" | "high" | "critical"
    recommendation: Optional[str] = None
    status: str = "pending"  # "pending" | "in_progress" | "completed" | "dismissed"

    def __repr__(self) -> str:
        return (
            f"ImpactIntervention(id={self.id}, severity={self.severity}, "
            f"topic={self.topic_id})"
        )

File: frontend/src/lib/types/impact.ts
/**
 * Impact Intelligence — TypeScript Types
 *
 * Type definitions for the Impact Intelligence module.
 */

// =========================================================================
// Base Types
// =========================================================================

export interface ImpactSchool {
  id: string
  name: string
  district?: string
  province?: string
  school_type?: 'primary' | 'secondary' | 'tertiary'
  active: boolean
  created: string
  updated: string
}

export interface ImpactClassGroup {
  id: string
  school_id: string
  name: string
  grade_level?: string
  academic_year?: string
  teacher_name?: string
  active: boolean
  created: string
  updated: string
}

export interface ImpactLearner {
  id: string
  school_id: string
  class_group_id: string
  learner_code: string
  display_name?: string
  status: 'active' | 'inactive' | 'transferred'
  created: string
  updated: string
}

export interface ImpactSubject {
  id: string
  name: string
  level?: string
  curriculum?: string
  created: string
  updated: string
}

export interface ImpactTopic {
  id: string
  subject_id: string
  name: string
  strand?: string
  syllabus_code?: string
  created: string
  updated: string
}

export interface ImpactAssessment {
  id: string
  school_id: string
  class_group_id: string
  subject_id: string
  title: string
  assessment_type: 'test' | 'exam' | 'quiz' | 'assignment'
  term?: string
  date_written?: string
  total_marks: number
  pass_mark?: number
  status: 'draft' | 'published' | 'graded'
  created: string
  updated: string
}

export interface ImpactAssessmentQuestion {
  id: string
  assessment_id: string
  question_number: number
  label?: string
  max_marks: number
  topic_id?: string
  skill_type: 'knowledge' | 'comprehension' | 'application' | 'analysis'
  difficulty?: 'easy' | 'medium' | 'hard'
  created: string
  updated: string
}

export interface ImpactMarkEntry {
  id: string
  assessment_id: string
  question_id: string
  learner_id: string
  score: number
  max_score: number
  created: string
  updated: string
}

export interface ImpactIntervention {
  id: string
  assessment_id: string
  class_group_id: string
  topic_id: string
  severity: 'low' | 'medium' | 'high' | 'critical'
  recommendation?: string
  status: 'pending' | 'in_progress' | 'completed' | 'dismissed'
  created: string
  updated: string
}

// =========================================================================
// Request Types
// =========================================================================

export interface ImpactSchoolCreate {
  name: string
  district?: string
  province?: string
  school_type?: 'primary' | 'secondary' | 'tertiary'
}

export interface ImpactSchoolUpdate {
  name?: string
  district?: string
  province?: string
  school_type?: 'primary' | 'secondary' | 'tertiary'
  active?: boolean
}

export interface ImpactClassGroupCreate {
  school_id: string
  name: string
  grade_level?: string
  academic_year?: string
  teacher_name?: string
}

export interface ImpactClassGroupUpdate {
  name?: string
  grade_level?: string
  academic_year?: string
  teacher_name?: string
  active?: boolean
}

export interface ImpactLearnerCreate {
  school_id: string
  class_group_id: string
  learner_code: string
  display_name?: string
  status?: 'active' | 'inactive' | 'transferred'
}

export interface ImpactLearnerUpdate {
  display_name?: string
  status?: 'active' | 'inactive' | 'transferred'
}

export interface ImpactSubjectCreate {
  name: string
  level?: string
  curriculum?: string
}

export interface ImpactSubjectUpdate {
  name?: string
  level?: string
  curriculum?: string
}

export interface ImpactTopicCreate {
  subject_id: string
  name: string
  strand?: string
  syllabus_code?: string
}

export interface ImpactTopicUpdate {
  name?: string
  strand?: string
  syllabus_code?: string
}

export interface ImpactAssessmentCreate {
  school_id: string
  class_group_id: string
  subject_id: string
  title: string
  assessment_type: 'test' | 'exam' | 'quiz' | 'assignment'
  term?: string
  date_written?: string
  total_marks: number
  pass_mark?: number
  status?: 'draft' | 'published' | 'graded'
}

export interface ImpactAssessmentUpdate {
  title?: string
  assessment_type?: 'test' | 'exam' | 'quiz' | 'assignment'
  term?: string
  date_written?: string
  total_marks?: number
  pass_mark?: number
  status?: 'draft' | 'published' | 'graded'
}

export interface ImpactAssessmentQuestionCreate {
  assessment_id: string
  question_number: number
  label?: string
  max_marks: number
  topic_id?: string
  skill_type: 'knowledge' | 'comprehension' | 'application' | 'analysis'
  difficulty?: 'easy' | 'medium' | 'hard'
}

export interface ImpactAssessmentQuestionUpdate {
  question_number?: number
  label?: string
  max_marks?: number
  topic_id?: string
  skill_type?: 'knowledge' | 'comprehension' | 'application' | 'analysis'
  difficulty?: 'easy' | 'medium' | 'hard'
}

export interface ImpactMarkEntryCreate {
  assessment_id: string
  question_id: string
  learner_id: string
  score: number
  max_score: number
}

export interface ImpactMarkEntryUpdate {
  score?: number
}

export interface ImpactInterventionCreate {
  assessment_id: string
  class_group_id: string
  topic_id: string
  severity: 'low' | 'medium' | 'high' | 'critical'
  recommendation?: string
  status?: 'pending' | 'in_progress' | 'completed' | 'dismissed'
}

export interface ImpactInterventionUpdate {
  severity?: 'low' | 'medium' | 'high' | 'critical'
  recommendation?: string
  status?: 'pending' | 'in_progress' | 'completed' | 'dismissed'
}

// =========================================================================
// Response Types
// =========================================================================

export interface ImpactSchoolListResponse {
  schools: ImpactSchool[]
  total: number
}

export interface ImpactClassGroupListResponse {
  class_groups: ImpactClassGroup[]
  total: number
}

export interface ImpactLearnerListResponse {
  learners: ImpactLearner[]
  total: number
}

export interface ImpactSubjectListResponse {
  subjects: ImpactSubject[]
  total: number
}

export interface ImpactTopicListResponse {
  topics: ImpactTopic[]
  total: number
}

export interface ImpactAssessmentListResponse {
  assessments: ImpactAssessment[]
  total: number
}

export interface ImpactAssessmentQuestionListResponse {
  questions: ImpactAssessmentQuestion[]
  total: number
}

export interface ImpactMarkEntryListResponse {
  mark_entries: ImpactMarkEntry[]
  total: number
}

export interface ImpactInterventionListResponse {
  interventions: ImpactIntervention[]
  total: number
}

// =========================================================================
// Analytics Types
// =========================================================================

export interface QuestionPerformance {
  question_id: string
  question_number: number
  label?: string
  max_marks: number
  topic_id?: string
  skill_type: string
  difficulty?: string
  total_score: number
  num_learners: number
  average_score: number
  average_percentage: number
  is_critical: boolean
}

export interface TopicPerformance {
  topic_id: string
  topic_name: string
  total_score: number
  total_max_marks: number
  percentage: number
  num_questions: number
  num_learners: number
  is_weak: boolean
  is_critical: boolean
}

export interface LearnerPerformance {
  learner_id: string
  learner_code: string
  display_name?: string
  total_score: number
  total_max_marks: number
  percentage: number
  passed: boolean
  risk_level: 'low' | 'medium' | 'high'
  questions_answered: number
  total_questions: number
}

export interface InterventionRecommendation {
  intervention_type: 'critical' | 'weak' | 'stable'
  entity_type: 'topic' | 'learner'
  entity_id: string
  entity_name: string
  severity: 'low' | 'medium' | 'high' | 'critical'
  recommendation: string
  percentage: number
}

export interface AssessmentAnalytics {
  assessment_id: string
  assessment_title: string
  assessment_type: string
  total_marks: number
  pass_mark?: number
  term?: string

  // Summary statistics
  total_learners: number
  learners_assessed: number
  mark_completion_rate: number
  class_average_percentage: number
  pass_rate: number
  failure_rate: number

  // Detailed performance
  question_performance: QuestionPerformance[]
  topic_performance: TopicPerformance[]
  learner_performance: LearnerPerformance[]

  // Weaknesses and interventions
  weak_topics: TopicPerformance[]
  at_risk_learners: LearnerPerformance[]
  interventions: InterventionRecommendation[]
}

// =========================================================================
// Dashboard Types
// =========================================================================

export interface SchoolDashboard {
  school_id: string
  school_name: string
  total_classes: number
  total_learners: number
  total_learners_assessed: number
  total_assessments: number
  overall_pass_rate: number
  pass_rate_by_subject: SubjectPassRate[]
  pass_rate_by_class: ClassPassRate[]
  weakest_topics: WeakTopic[]
  classes_needing_support: ClassNeedingSupport[]
  recent_interventions: RecentIntervention[]
}

export interface SubjectPassRate {
  subject_id: string
  subject_name: string
  total_learners: number
  pass_rate: number
}

export interface ClassPassRate {
  class_id: string
  class_name: string
  total_learners: number
  pass_rate: number
}

export interface WeakTopic {
  topic_id: string
  topic_name: string
  percentage: number
  is_critical: boolean
  num_questions: number
}

export interface ClassNeedingSupport {
  class_id: string
  class_name: string
  pass_rate: number
  total_learners: number
}

export interface RecentIntervention {
  id: string
  severity: 'low' | 'medium' | 'high' | 'critical'
  recommendation?: string
  status: 'pending' | 'in_progress' | 'completed' | 'dismissed'
  created: string
}

export interface MinistryDashboard {
  total_schools: number
  total_learners: number
  total_learners_assessed: number
  total_assessments: number
  average_pass_rate: number
  weak_topics_by_subject: SubjectWeakTopics[]
  schools_needing_support: SchoolNeedingSupport[]
  classes_needing_support: ClassNeedingSupportMinistry[]
}

export interface SubjectWeakTopics {
  subject_id: string
  subject_name: string
  weak_topics: WeakTopicMinistry[]
}

export interface WeakTopicMinistry {
  topic_name: string
  percentage: number
  is_critical: boolean
}

export interface SchoolNeedingSupport {
  school_id: string
  school_name: string
  pass_rate: number
  total_assessments: number
}

export interface ClassNeedingSupportMinistry {
  class_id: string
  class_name: string
  school_id: string
  pass_rate: number
  total_learners: number
}

// =========================================================================
// Report Types
// =========================================================================

export interface AssessmentReport {
  assessment: ImpactAssessment
  questions: ImpactAssessmentQuestion[]
  learners: ImpactLearner[]
  analytics: AssessmentAnalytics
  school?: ImpactSchool
  class_group?: ImpactClassGroup
  subject?: ImpactSubject
}

export interface SchoolReport {
  school: ImpactSchool
  classes: ImpactClassGroup[]
  assessments: ImpactAssessment[]
  pass_rate_by_class: ClassPassRate[]
  recent_interventions: RecentIntervention[]
  total_learners: number
  total_learners_assessed: number
  overall_pass_rate: number
  subjects: ImpactSubject[]
  weak_topics: WeakTopic[]
  support_indicators: {
    learners_needing_support: number
    classes_needing_support: number
    open_interventions: number
  }
  data_quality: {
    status: 'ready' | 'review'
    mark_completion_rate: number
    mapped_question_rate: number
  }
  disclosure: string
  limitations: string[]
}

// =========================================================================
// AI Summary Types
// =========================================================================

export interface TeacherSummary {
  summary: string
  revision_sequence: string
  source: 'ai-generated' | 'fallback'
  error?: string
}

export interface InterventionPlan {
  plan: string
  source: 'ai-generated' | 'fallback'
  error?: string
}

export interface RemedialLesson {
  outline: string
  mini_test_idea: string
  source: 'ai-generated' | 'fallback'
  error?: string
}
